internal/server/run.go

```diff
@@ -0,0 +1,93 @@
+package server
+
+import (
+	"errors"
+	"io"
+	"log"
+	"net"
+	"tcp-load-balancer/internal/upstream"
+)
+
+var ErrUninitialized = errors.New("load balancer not initialized")
+
+// Run handles incoming connections until terminated.
+func (l *LoadBalancer) Run() error {
+	if l == nil || l.listener == nil {
+		return ErrUninitialized
+	}
+
+	for {
+		// TODO: Set up mTLS in next PR. For now, connect without TLS.
+		clientConn, err := l.listener.Accept()
+		if err != nil {
+			go respondAndClose(clientConn, err.Error())
+			continue
+		}
+
+		if err := l.handleConnection(clientConn); err != nil {
+			log.Printf("Unable to handle connection: %s", err)
+			go respondAndClose(clientConn, err.Error())
+		}
+	}
+}
+
+// handleConnection selects an upstream host, tracks connection counts, and forwards data upstream.
+func (l *LoadBalancer) handleConnection(clientConn net.Conn) error {
```

Comment:
Does the clientConn.Close function is called anywhere ? If not it missing Close call will lead to resource leak.

Reply:
It was not handled. I just added the `defer clientConn.Close()` during setup (within Run). That will be handled now.